### PR TITLE
check status of router, registry, metrics, logging, imagestreams in oc cluster status

### LINKED ISSUE #12224: oc cluster status extensions
oc cluster up status should check the status of the router, registry, metrics (if applicable), logging (if applicable)

This is because all of these may continue loading after oc cluster up itself has finished.

Possibly some sort of option equivalent to --porcelain for checking via shell scripts may be appropriate.

@csrwng fyi

### PR DESCRIPTION
**Pull request by jim-minter**
fixes #12224

@openshift/devex 
@csrwng

### CODE REVIEW

**Comment by csrwng** [comment]
@jim-minter just a general comment... not sure that a warning is the best way to report that the registry/router/etc. is not yet running. It may lead people to think that there's something wrong with their cluster. I would simply report that it's either running or not running. Maybe a warning is warranted only when the uptime of the container exceeds a certain threshold and the registry or router are still not functional.

File: pkg/bootstrap/docker/status.go [hunk 1]
```diff
@@ -94,6 +97,56 @@ func (c *ClientStatusConfig) Status(f *clientcmd.Factory, out io.Writer) error {
...
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: true,
+				},
+			},
+		}
+
+		resp, _ := insecureCli.Get(config.AssetConfig.MetricsPublicURL + "/status")
+		if resp == nil || resp.StatusCode != http.StatusOK {
+			fmt.Println("Warning: metrics are not yet successfully deployed")
```
Comment (csrwng) [inline pkg/bootstrap/docker/status.go]: s /are/is/ ... I would tend to think of metrics as a singular component

**Comment by csrwng** [comment]
It would also be helpful to report the status of the persistent volume job (-n default job/persistent-volume-setup)

**Comment by jim-minter** [comment]
@csrwng all done

**Comment by csrwng** [comment]
So I tried this locally ... I get this output:

```
The OpenShift cluster was started 31 seconds ago

Web console URL: https://127.0.0.1:8443

Config is at host directory /var/lib/origin/openshift.local.config
Volumes are at host directory /var/lib/origin/openshift.local.volumes
Persistent volumes are at host directory /var/lib/origin/openshift.local.pv
Data will be discarded when cluster is destroyed

Router is not yet successfully deployed
Persistent volumes are not yet successfully created
Error: OpenShift cluster additional health checks failed
```

So a couple of nits ... 

1) the message about persistent volumes could be better... maybe say something to the effect that creation has not completed. Even when the job hasn't completed, you are able to start using PV's since some will have been created.

2) The last line would sound better if it said something like provisioning tasks are still in progress... which is not necessarily an error.

Also, just a question, why are we doing the logging check that way? (Instead of for example hitting a URL like we do with metrics?)

**Comment by jim-minter** [comment]
> Also, just a question, why are we doing the logging check that way? (Instead of for example hitting a URL like we do with metrics?)

Because of the oauth proxy, I couldn't find a suitable URL to hit.  If you can find one, do let me know!

**Comment by csrwng** [comment]
@sosiouxme @jcantrill please see above. Is there a reliable check that we can do to ensure that logging is working?

**Comment by sosiouxme** [comment]
"Working" comprises a lot of different things, some more subtle than others. But if we're just looking for "present" then why not hit the public logging URL and check that it returns a 302 status (as opposed to 503 if nothing were listening)? Technically the logging stack doesn't have to be deployed in the `logging` namespace; and I don't think you want to recreate `oc adm diagnostics AggregatedLogging` in order to get a thorough test for "working".

**Comment by jim-minter** [comment]
@csrwng updated and repushed:

```
[root@default origin]# oc cluster status
The OpenShift cluster was started 15 seconds ago

Web console URL: https://127.0.0.1:8443

Config is at host directory /var/lib/origin/openshift.local.config
Volumes are at host directory /var/lib/origin/openshift.local.volumes
Persistent volumes are at host directory /var/lib/origin/openshift.local.pv
Data will be discarded when cluster is destroyed

Notice: Docker registry is not yet ready
Notice: Router is not yet ready
Notice: Persistent volumes are not yet ready
Notice: Imagestreams are not yet ready

Notice: 4 OpenShift component(s) are not yet ready (see above)
```

**Comment by csrwng** [comment]
LGTM
[merge]

**Comment by jim-minter** [comment]
re-pushed with fix to failing 'oc cluster status' test, ptal/remerge

**Comment by csrwng** [comment]
[merge]

File: pkg/bootstrap/docker/status.go [hunk 2]
```diff
@@ -92,7 +97,80 @@ func (c *ClientStatusConfig) Status(f *clientcmd.Factory, out io.Writer) error {
...
+
+	eh := exec.NewExecHelper(dockerClient, openshift.OpenShiftContainer)
+
+	stdout, _, _ := eh.Command(strings.Split("oc get dc docker-registry -n default -o template --template {{.status.availableReplicas}}", " ")...).Output()
+	if stdout != "1" {
+		fmt.Fprintln(out, "Notice: Docker registry is not yet ready")
+		notReady++
+	}
+
+	stdout, _, _ = eh.Command(strings.Split("oc get dc router -n default -o template --template {{.status.availableReplicas}}", " ")...).Output()
```
Comment (smarterclayton) [inline pkg/bootstrap/docker/status.go]: Why are you using split here and not below?  Don't use split.

File: pkg/bootstrap/docker/status.go [hunk 3]
```diff
@@ -92,7 +97,80 @@ func (c *ClientStatusConfig) Status(f *clientcmd.Factory, out io.Writer) error {
...
 	}
 
-	fmt.Print(status(container, config))
+	fmt.Fprint(out, status(container, config))
+
+	notReady := 0
+
+	eh := exec.NewExecHelper(dockerClient, openshift.OpenShiftContainer)
+
+	stdout, _, _ := eh.Command(strings.Split("oc get dc docker-registry -n default -o template --template {{.status.availableReplicas}}", " ")...).Output()
```
Comment (smarterclayton) [inline pkg/bootstrap/docker/status.go]: Don't use split.

Comment (kargakis) [inline pkg/bootstrap/docker/status.go]: What if the registry runs as a daemon set?

Comment (kargakis) [inline pkg/bootstrap/docker/status.go]: I guess this is based on the default set by `oc cluster up`.

**Comment by jim-minter** [comment]
@csrwng repushed to take care of @smarterclayton's comments above - please remerge

**Comment by jim-minter** [comment]
@csrwng ping :)

**Comment by csrwng** [comment]
[merge]

**Comment by csrwng** [comment]
Flakes https://github.com/openshift/origin/issues/14385 and https://github.com/openshift/origin/issues/14590
[merge]

**Comment by csrwng** [comment]
https://github.com/openshift/origin/issues/14603
https://github.com/openshift/origin/issues/14208
[merge]

**Comment by csrwng** [comment]
https://github.com/openshift/origin/issues/14523
[merge]

**Pull request merged**